### ISSUE TITLE
IPv6 Compliance RFC7217: Resolving DAD Conflicts (Host Only) [v6LC.3.2.7 Part B]

### ISSUE THREAD

**Issue by LiveFreeAndRoam** [systemd/systemd#31605]
### systemd version the issue has been seen with

255

### Used distribution

Ubuntu 23

### Linux kernel version used

6.5.0-17-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

#### v6LC.3.2.7: Resolving DAD Conflicts (Host Only) - Part B: Global
**Purpose**: Verify that the HUT keeps the network interface constant across system network events.
**Reference**: [RFC7217 – Section 6](https://datatracker.ietf.org/doc/html/rfc7217#section-6)

#### Network Configuration
```
       TR1
        |
    +---+---+
    |       |
   TN1     HUT
```

#### Procedure
| Step | Action | Expected Behavior of HUT<sup>(1)</sup> |
|--------|--------|--------|
| 1 | Initialize the interface on the HUT. | |
| 2 | Allow time for the HUT to perform stateless address autoconfiguration and DAD. | The HUT should perform DAD on its tentative address for its interface by sending DAD Neighbor Solicitations. The HUT should assign the tentative address to its interface. Interface IDs are required to be 64 bits long and use the algorithm from RFC 7217. |
| 3 | TR1 transmits a Router Advertisement A on Link A. | |
| 4 | Allow time for the DUT to perform stateless address autoconfiguration and DAD. | The HUT should perform DAD on its tentative global address for its interface on Link A. Interface IDs are required to be 64 bits long and use the algorithm from RFC 7217. The Interface ID should be different then the id used in Step 2. |
| 5 | After TN1 receives a DAD NS message from the NUT. TN1 transmits DAD Neighbor Advertisement with the Target Address set to the NUT’s tentative global address. | |
| 6 | Wait IDGEN_DELAY (1 second) | The HUT should transmit a DAD NS message with the Target Address set to a different tentative global address then in Step 2. |
* **(1)** - Host Under Test

### Unexpected behaviour you saw

In Step 6, the HUT should have seen that its tentative global address was already owned by TN1 and so transmitted a DAD NS with a new Target Address. In the case of Networkd, it did not send a new DAD NS.

From the log, below, it appears that Networkd does not assign the address, so it at least avoids duplicate addresses.  However, it leaves the interface without a global address; it ought to have tried a new global address.

### Steps to reproduce the problem

[v6LC_3_2_07b.zip](https://github.com/systemd/systemd/files/14477483/v6LC_3_2_07b.zip)

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
enp0s9: Flags change: -UP -LOWER_UP -RUNNING
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=30 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: Link DOWN
enp0s9: Lost carrier
enp0s9: DHCPv6 lease lost
enp0s9: NDISC: Stopping IPv6 Router Solicitation client
enp0s9: Forgetting removed foreign route (n/a): dst: fe80::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 256, proto: kernel, scope: global, type: unicast, flags: n/a
enp0s9: Forgetting removed foreign route (n/a): dst: fe80::7d36:7adb:773f:aa0b/128, src: n/a, gw: n/a, prefsrc: n/a, table: local(255), priority: 0, proto: kernel, scope: global, type: local, flags: n/a
enp0s9: Forgetting removed foreign route (n/a): dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a, table: local(255), priority: 256, proto: kernel, scope: global, type: multicast, flags: n/a
enp0s9: Forgetting removed foreign address (n/a): fe80::7d36:7adb:773f:aa0b/64 (valid forever, preferred forever), flags: permanent,stable-privacy, scope: link
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=31 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: Flags change: +UP
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=32 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: Link UP
enp0s9: Received new foreign route (configured): dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a, table: local(255), priority: 256, proto: kernel, scope: global, type: multicast, flags: n/a
enp0s9: Received new foreign route (configured): dst: fe80::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 256, proto: kernel, scope: global, type: unicast, flags: n/a
enp0s9: Flags change: +LOWER_UP +RUNNING
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=33 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: Gained carrier
enp0s9: found matching network '/etc/systemd/network/10-enp0s9.network'.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: link_check_ready(): IPv6LL is not configured yet.
enp0s9: Received new foreign address (configured): fe80::7d36:7adb:773f:aa0b/64 (valid forever, preferred forever), flags: permanent,stable-privacy, scope: link
enp0s9: Gained IPv6LL
enp0s9: Discovering IPv6 routers
enp0s9: NDISC: Started IPv6 Router Solicitation client
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_34 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=34 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Received new foreign route (configured): dst: fe80::7d36:7adb:773f:aa0b/128, src: n/a, gw: n/a, prefsrc: n/a, table: local(255), priority: 0, proto: kernel, scope: global, type: local, flags: n/a
enp0s9: NDISC: Sent Router Solicitation, next solicitation in 3s
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 2h 30min
enp0s9: NDISC: Invoking callback for 'router' event.
Setting '/proc/sys/net/ipv6/neigh/enp0s9/base_reachable_time_ms' to '30000'
No change in value '30000', suppressing write
Setting '/proc/sys/net/ipv6/neigh/enp0s9/retrans_time_ms' to '1000'
No change in value '1000', suppressing write
Setting '/proc/sys/net/ipv6/conf/enp0s9/hop_limit' to '64'
No change in value '64', suppressing write
enp0s9: Requesting NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a

#
# DAD NS, who owns 2001:2:0:1000:f1c5:7f81:2ff6:19fd/64?
#
enp0s9: Requesting NDisc address (n/a): 2001:2:0:1000:f1c5:7f81:2ff6:19fd/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Setting SLAAC addresses and router.
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Configuring NDisc route (requesting): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Configuring NDisc route (requesting): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a

#
# Configuring 2001:2:0:1000:f1c5:7f81:2ff6:19fd/64.
#
enp0s9: Configuring NDisc address (requesting): 2001:2:0:1000:f1c5:7f81:2ff6:19fd/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: manage-temporary-address,no-prefixroute, scope: global
enp0s9: Received new NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.
enp0s9: Received new NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: ndisc_check_ready(): SLAAC addresses and routes are not set.

#
# Is this where TN1 claims ownership of 2001:2:0:1000:f1c5:7f81:2ff6:19fd/64?
#
enp0s9: Received new NDisc address (configured): 2001:2:0:1000:f1c5:7f81:2ff6:19fd/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: tentative,manage-temporary-address,no-prefixroute, scope: global
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: ndisc_check_ready(): no SLAAC address is ready.
ICMPv6: Received a packet from neither link-local nor null address.
enp0s9: NDISC: Failed to receive ICMPv6 packet, ignoring: Cannot assign requested address

#
# Networkd drops 2001:2:0:1000:f1c5:7f81:2ff6:19fd/64.
#
enp0s9: Forgetting removed NDisc address (n/a): 2001:2:0:1000:f1c5:7f81:2ff6:19fd/64 (valid for 18h 12min 14s, preferred for 18h 12min 14s), flags: tentative,manage-temporary-address,no-prefixroute, scope: global
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.

#
# RA cleanup prefixes, set lifetimes to 0.
#
enp0s9: NDISC: Received Router Advertisement: flags none preference medium lifetime 0
enp0s9: NDISC: Invoking callback for 'router' event.
enp0s9: Removing NDisc route (configured): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Found an existing route that conflicts with new route based on a received RA, removing.
enp0s9: Removing NDisc route (configured): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Requesting NDisc route (configured,removing): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Setting SLAAC addresses and router.
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
enp0s9: Refuse to configure NDisc route with zero lifetime.
enp0s9: Forgetting removed NDisc route (n/a): dst: n/a, src: n/a, gw: fe80::200:10ff:fe10:1060, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: Forgetting removed NDisc route (n/a): dst: 2001:2:0:1000::/64, src: n/a, gw: n/a, prefsrc: n/a, table: main(254), priority: 1024, proto: ra, scope: global, type: unicast, flags: n/a
enp0s9: NDISC: Received Neighbor Advertisement from fe80::200:10ff:fe10:1060: Router=yes, Solicited=no, Override=yes
enp0s9: NDISC: Invoking callback for 'neighbor' event.
enp0s9: NDISC: No RA received before link confirmation timeout
enp0s9: NDISC: Invoking callback for 'timeout' event.
enp0s9: NDisc handler get timeout event
enp0s9: link_check_ready(): dynamic addressing protocols are enabled but none of them finished yet.
```

**Comment by yuwata** [comment]
Fix is waiting in #32166 (the same commits are included in #30569).

**Comment by LiveFreeAndRoam** [comment]
Repeating what I said in #32166...

Great work @yuwata!! The test now passes. I have attached the PCAP showing the address collission and the HUT restarting DAD with a modified GA.

[v6LC_3_2_07b-networkd-pass.zip](https://github.com/systemd/systemd/files/14913550/v6LC_3_2_07b-networkd-pass.zip)

This was tested with:

```
 git log -n 1
commit e680ebf01ef0c6d6f1feddea77a27
```

BTW: This just leaves 4 failing cases (after starting with 105 failing cases).  I must say, a remarkable effort @yuwata.
